store/pebble/vk_merger.go

```diff
@@ -98,21 +104,59 @@ func (v *valueKeysValueMerger) DeletableFinish(includesBase bool) ([]byte, bool,
 	return b, len(b) == 0, c, err
 }
 
-func (v *valueKeysValueMerger) deleteValueKey(value []byte) {
-	b := v.merge[:0]
-	for _, x := range v.merge {
-		if !bytes.Equal(x, value) {
-			b = append(b, x)
+// prune removes value-keys that are present in deletes from merges
+func (v *valueKeysValueMerger) prune() {
+	pruned := v.merges[:0]
+	for _, x := range v.merges {
+		if _, ok := v.deletes[string(x)]; !ok {
+			pruned = append(pruned, x)
 		}
 	}
-	v.merge = b
+	v.merges = pruned
 }
 
+// addToMerges checks whether the given value exists and if not adds it to the list of merges.
+func (v *valueKeysValueMerger) addToMerges(value []byte) {
+	if !v.exists(value) {
+		dst := make([]byte, len(value))
+		copy(dst, value)
+		v.merges = append(v.merges, dst)
+	}
+}
+
+// exists checks whether the given value is already present, either pending merge or deletion.
 func (v *valueKeysValueMerger) exists(value []byte) bool {
-	for _, x := range v.merge {
+	if _, pendingDelete := v.deletes[string(value)]; pendingDelete {
+		return true
+	}
+	for _, x := range v.merges {
```

Comment:
How many elements do you expect to be in `v.merges`? Would it  make sense to have an additional hashtable to avoid full scan?

Reply:
The reason for not using a `map` to store `merges` is that we need to maintain the insertion order. Iteration order over `map` Golang is non-deterministic I'm afraid.